File: Story4/TheMazeRunner.java
/**
 
 * Description- A mouse is in a $N \times N$ grid maze. It starts at $(0,0)$ and needs 
to reach $(N-1, N-1)$. The mouse can only move Down or Right. Some 
cells are blocked (0) and some are open (1). Find if a path exists.
 * Author - Naval
 */
import java.util.Scanner;

public class TheMazeRunner {

    public static boolean isPathExists(int[][] maze, int n) {
        return backtrack(maze, 0, 0, n);
    }

    private static boolean backtrack(int[][] maze, int row, int col, int n) {
        if (row >= n || col >= n || maze[row][col] == 0) {
            return false;
        }
        if (row == n - 1 && col == n - 1) {
            return true;
        }
        return backtrack(maze, row, col + 1, n) ||
                backtrack(maze, row + 1, col, n);
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter maze size (N): ");
        int n = sc.nextInt();
        int[][] maze = new int[n][n];
        System.out.println("Enter maze (0 = blocked, 1 = open):");
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                maze[i][j] = sc.nextInt();
            }
        }


        boolean result = isPathExists(maze, n);
        System.out.println("Path exists: " + result);

        sc.close();
    }
}

File: week4_story/TheMazeRunner.cpp


/**
 * @filename   : TheMazeRunner.cpp
 * @description: A mouse is in a times N grid maze. It starts at $(0,0)$ and needs to reach (N-1, N-1).
 *  The mouse can only move Down or Right. Some cells are blocked (0) and some are open (1). Find if a path exists using backtracking.
 * @author     : Ehtesham Nawaz
 */

// Input: [[1, 0], [1, 1]]
// Output: True

#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> cache;
// added caching to bring exponential time complexit from O(2^n*n) to O(n*n);
bool isPathExists(int currentRow, int currentCol, vector<vector<int>> &grid)
{
    if (currentRow < 0 || currentCol < 0)
        return false;
    if (grid[currentRow][currentCol] == 0)
        return false;
    if (currentRow == 0 && currentCol == 0)
        return true;

    if (cache[currentRow][currentCol] != -1)
        return cache[currentRow][currentCol];

    int moveUp = isPathExists(currentRow - 1, currentCol, grid);
    int moveLeft = isPathExists(currentRow, currentCol - 1, grid);
    return cache[currentRow][currentCol] = moveLeft || moveUp;
}

int main()
{
    int sizeOfGrid = 2;
    vector<vector<int>> grid = {{1, 0}, {1, 1}};

    cache.resize(sizeOfGrid, vector<int>(sizeOfGrid, -1));

    cout << (isPathExists(1, 1, grid) ? "True" : "False");

    return 0;
}
